pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1459,6 +1460,29 @@ private ByteBuf processMessageChunk(ByteBuf compressedPayload, MessageMetadata m
         // discard message if chunk is out-of-order
         if (chunkedMsgCtx == null || chunkedMsgCtx.chunkedMsgBuffer == null
                 || msgMetadata.getChunkId() != (chunkedMsgCtx.lastChunkedMessageId + 1)) {
+            // Filter and ack duplicated chunks instead of discard ctx.
+            // For example:
+            //     Chunk-1 sequence ID: 0, chunk ID: 0, msgID: 1:1
+            //     Chunk-2 sequence ID: 0, chunk ID: 1, msgID: 1:2
+            //     Chunk-3 sequence ID: 0, chunk ID: 2, msgID: 1:3
+            //     Chunk-4 sequence ID: 0, chunk ID: 1, msgID: 1:4
+            //     Chunk-5 sequence ID: 0, chunk ID: 2, msgID: 1:5
+            //     Chunk-6 sequence ID: 0, chunk ID: 3, msgID: 1:6
+            // We should filter and ack chunk-4 and chunk-5.
```

Comment:
It could be risky here if the producer side changed the chunk policy. For example:

```
Chunk-1 sequence ID: 0, chunk ID: 0, size 10kb
Chunk-1 sequence ID: 0, chunk ID: 1, size 10kb
Chunk-1 sequence ID: 0, chunk ID: 2, size 10kb

```

The producer restarted and sent the message with the same seq with the changed chunk policy.

```
Chunk-1 sequence ID: 0, chunk ID: 0, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 1, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 2, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 3, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 4, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 5, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 6, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 7, size 5kb
```

If you filter out 

```
Chunk-1 sequence ID: 0, chunk ID: 0, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 1, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 2, size 5kb
```

You will get a corrupted message:

```
Chunk-1 sequence ID: 0, chunk ID: 0, size 10kb
Chunk-1 sequence ID: 0, chunk ID: 1, size 10kb
Chunk-1 sequence ID: 0, chunk ID: 2, size 10kb
Chunk-1 sequence ID: 0, chunk ID: 3, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 4, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 5, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 6, size 5kb
Chunk-1 sequence ID: 0, chunk ID: 7, size 5kb
```

IMO, the correct solution is always to discard the unsealed chunks and make sure the producer side should only resend the whole chunk message instead of part of the chunks.

For example:

```
            //     Chunk-1 sequence ID: 0, chunk ID: 0, msgID: 1:1
            //     Chunk-2 sequence ID: 0, chunk ID: 1, msgID: 1:2
            //     Chunk-3 sequence ID: 0, chunk ID: 2, msgID: 1:3
            //     Chunk-4 sequence ID: 0, chunk ID: 1, msgID: 1:4
            //     Chunk-5 sequence ID: 0, chunk ID: 2, msgID: 1:5
            //     Chunk-6 sequence ID: 0, chunk ID: 3, msgID: 1:6
```

It shouldn't happen. You should always see

```
            //     Chunk-1 sequence ID: 0, chunk ID: 0, msgID: 1:1
            //     Chunk-2 sequence ID: 0, chunk ID: 1, msgID: 1:2
            //     Chunk-3 sequence ID: 0, chunk ID: 2, msgID: 1:3
            //     Chunk-4 sequence ID: 0, chunk ID: 0, msgID: 1:4
            //     Chunk-5 sequence ID: 0, chunk ID: 1, msgID: 1:5
            //     Chunk-6 sequence ID: 0, chunk ID: 2, msgID: 1:6
            //     Chunk-7 sequence ID: 0, chunk ID: 3, msgID: 1:7
```

Reply:
Hi @codelipenghui, This case is handled by the https://github.com/apache/pulsar/pull/21070

Reply:
Oh, I got it. It looks good to me.